### PR TITLE
Add a `ResourcesOperation` class to support resource estimation of custom operations

### PR DESCRIPTION
**Pull request by Jaybsoni**
### Before submitting

Please complete the following checklist when submitting a PR:

- [ ] All new features must include a unit test.
      If you've fixed a bug or added code that should be tested, add a test to the
      test directory!

- [ ] All new functions and code must be clearly commented and documented.
      If you do make documentation changes, make sure that the docs build and
      render correctly by running `make docs`.

- [ ] Ensure that the test suite passes, by running `make test`.

- [ ] Add a new entry to the `doc/releases/changelog-dev.md` file, summarizing the
      change, and including a link back to the PR.

- [ ] The PennyLane source code conforms to
      [PEP8 standards](https://www.python.org/dev/peps/pep-0008/).
      We check all of our code against [Pylint](https://www.pylint.org/).
      To lint modified files, simply `pip install pylint`, and then
      run `pylint pennylane/path/to/file.py`.

When all the above are checked, delete everything above the dashed
line and fill in the pull request template.

------------------------------------------------------------------------------------------------------------

**Context:**

**Description of the Change:**

**Benefits:**

**Possible Drawbacks:**

**Related GitHub Issues:**

### CODE REVIEW

File: pennylane/resource/resource.py [hunk 1]
```diff
@@ -71,6 +75,40 @@ def _ipython_display_(self):
         print(str(self))
 
 
+class ResourcesOperation(Operation):
+    r"""Base class representing quantum gates or channels applied to quantum states
+    storing additionally, the resource requirements of the quantum gate.
```
Comment (DSGuala) [inline pennylane/resource/resource.py:80]:
```suggestion
    r"""Base class that represents quantum gates or channels applied to quantum 
    states and stores the resource requirements of the quantum gate.
```

File: tests/resource/test_resource.py [hunk 2]
```diff
@@ -131,13 +131,50 @@ def test_ipython_display(self, r, rep, capsys):
         assert rep in captured.out
 
 
+class TestResourcesOperation:
+    """Test that the ResourcesOperation class is constructed correctly"""
+
+    def test_raise_not_implemented_error(self):
```
Comment (DSGuala) [inline tests/resource/test_resource.py:137]: Nice :+1:

File: pennylane/resource/resource.py [hunk 3]
```diff
@@ -88,7 +126,14 @@ def _count_resources(tape, shots: int) -> Resources:
...
     for op in tape.operations:
-        gate_types[op.name] += 1
-        num_gates += 1
+        if isinstance(op, ResourcesOperation):
+             op_resource = op.resources()
+             for d in op_resource.gate_types:
+                 gate_types[d] += op_resource.gate_types[d]
+             num_gates += sum(op_resource.gate_types.values())
+
+         else:
```
Comment (DSGuala) [inline pennylane/resource/resource.py:135]:
```suggestion
        else:
```
Slightly out of alignment leading to an error.

**Review by DSGuala — APPROVED** [review]
Small suggestions. Approved so that merging is not blocked. Pending docstrings and changelog updates.

File: pennylane/resource/resource.py [hunk 4]
```diff
@@ -71,6 +75,40 @@ def _ipython_display_(self):
...
+    @abstractmethod
+    def resources(self) -> Resources:
+        r"""Compute the resources required for this operation.
+
+        Returns:
+            (:class:`~.Resources`): The resources required by this operation.
+
+        **Examples**
+
+        >>> class CustomOp(OperationWithResources):
```
Comment (Jaybsoni) [inline pennylane/resource/resource.py:96]:
```suggestion
        >>> class CustomOp(ResourcesOperation):
```

**Review by soranjh — APPROVED** [review]

File: doc/releases/changelog-dev.md [hunk 5]
```diff
@@ -256,6 +256,39 @@
...
+  ```pycon
+  >>> class CustomOp(ResourcesOperation):
+  ...     def resources(self):
+  ...         return Resources(num_wires=1, num_gates=2, gate_types={"PauliX": 2})
+  ... 
+  >>> CustomOp(wires=1)
+  CustomOp(wires=[1])
+  ```
+  
+  Then we can track and display the resources of the workflow using `tape.specs`:
```
Comment (soranjh) [inline doc/releases/changelog-dev.md:273]: Minor point: should we use an example with `qml.specs`?

File: pennylane/resource/resource.py [hunk 6]
```diff
@@ -88,7 +126,18 @@ def _count_resources(tape, shots: int) -> Resources:
     num_gates = 0
     gate_types = defaultdict(int)
     for op in tape.operations:
-        gate_types[op.name] += 1
-        num_gates += 1
+        if isinstance(op, ResourcesOperation):
+            op_resource = op.resources()
+
+            if op_resource.depth > 1:
+                depth = None  # Cannot be determined with custom depth operations
```
Comment (soranjh) [inline pennylane/resource/resource.py:133]: Nice.

File: doc/releases/changelog-dev.md (hunk 5 above)
Comment (Jaybsoni) [reply doc/releases/changelog-dev.md]: Added 👍🏼

File: doc/releases/changelog-dev.md [hunk 7]
```diff
@@ -256,6 +256,39 @@
...
 
+* The `ResourcesOperation` class is added to allow users to define operations with custom resource information.
+  [(#4026)](https://github.com/PennyLaneAI/pennylane/pull/4026)
+
+  For example, users can define a custom operation by inheriting from this new class:
+
+  ```pycon
+  >>> class CustomOp(ResourcesOperation):
+  ...     def resources(self):
+  ...         return Resources(num_wires=1, num_gates=2, gate_types={"PauliX": 2})
```
Comment (soranjh) [inline doc/releases/changelog-dev.md:267]: Better to use `num_wires=2` to be compatible with depth of 1. Might also define `depth` here.

File: doc/releases/changelog-dev.md [hunk 8]
```diff
@@ -256,6 +256,39 @@
...
+  >>> @qml.qnode(dev)
+  ... def circ():
+  ...     qml.PauliZ(wires=0)
+  ...     CustomOp(wires=1)
+  ...     return qml.state()
+  ... 
+  >>> print(qml.specs(circ)()['resources'])
+  wires: 2
+  gates: 3
+  depth: 1
```
Comment (soranjh) [inline doc/releases/changelog-dev.md:286]: This should be 2?

File: pennylane/resource/resource.py [hunk 9]
```diff
@@ -71,6 +75,40 @@ def _ipython_display_(self):
         print(str(self))
 
 
+class ResourcesOperation(Operation):
+    r"""Base class that represents quantum gates or channels applied to quantum 
+    states and stores the resource requirements of the quantum gate.
+
+    .. note::
+        Child classes must implement the :func:`resources()` method which computes
+        the resource requirements of the operation.
```
Comment (soranjh) [inline pennylane/resource/resource.py:84]: The same example in changelog might be added here.

File: tests/resource/test_resource.py [hunk 10]
```diff
@@ -131,13 +131,50 @@ def test_ipython_display(self, r, rep, capsys):
...
+        class CustomOPWithResources(ResourcesOperation):
+            num_wires = 2
+
+            def resources(self):
+                return Resources(num_wires=self.num_wires)
+
+        with pytest.raises(TypeError, match="Can't instantiate"):
+            _ = CustomOpNoResource()
+
+        assert CustomOPWithResources  # shouldn't raise an error
```
Comment (trbromley) [inline tests/resource/test_resource.py:153]: What does this line do? Do we need to instantiate `CustomOpWithResources`?

**Review by trbromley — APPROVED** [review]
Thanks @Jaybsoni!

File: pennylane/resource/resource.py [hunk 11]
```diff
@@ -71,6 +75,40 @@ def _ipython_display_(self):
...
+            (:class:`~.Resources`): The resources required by this operation.
+
+        **Examples**
+
+        >>> class CustomOp(ResourcesOperation):
+        ...     num_wires = 2
+        ...     def resources(self):
+        ...         return Resources(num_wires=self.num_wires, num_gates=3, depth=2)
+        ...
+        >>> op = CustomOp()
```
[the remark this answers is not in the document]
Comment (Jaybsoni) [reply pennylane/resource/resource.py:101]: Added 👍🏼

**Comment by Jaybsoni** [comment]
[sc-37600]

**Pull request merged**